torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -2247,88 +2385,220 @@ def _load_sharded_state_dict(
+        """
+        Moves the :meth:`forward` inputs to the compute device and casts them
+        to the appropriate dtype if needed.
+        """
+        if not self._is_root:
+            return
+        # TODO: Do not use the side stream for tensor copies for now;
+        # investigate the perf with/without it
+        # TODO: For mixed precision, move the inputs to the compute device and
+        # cast to reduced-precision in a single `to()` call
+        args, kwargs = _to_kwargs(args, kwargs, self.compute_device.index, False)
+        args = args[0]
+        kwargs = kwargs[0]
+        if self._mixed_precision_enabled_for_params():
+            input_dtype = self.mixed_precision.param_dtype
+            args, kwargs = self._cast_fp_inputs_to_precision(
+                input_dtype, *args, **kwargs,
+            )
+
+    def forward(self, *args: Any, **kwargs: Any) -> Any:
+        """
+        Runs the forward pass. For the recursive-wrapping path, the pre-forward
+        and post-forward are called explicitly, while for the non-recursive-
+        wrapping path, they are registered as hooks on every (sub)module.
+        """
+        with torch.autograd.profiler.record_function("FullyShardedDataParallel.forward"):    
+            # Non-recursive wrapping path (using hooks)
+            if self._use_param_exec_order_policy:
+                self._cast_forward_inputs(*args, **kwargs)
+                return self._fsdp_wrapped_module(*args, **kwargs)
+            # Recursive wrapping path
+            unused = None
+            free_full_params = self.reshard_after_forward
+            free_mp_shard = (
+                self.reshard_after_forward and
+                self._mixed_precision_enabled_for_params()
+            )
+            reshard_fn = functools.partial(
+                self._reshard,
+                [handle.flat_param for handle in self._handles],
+                free_full_params,
+                free_mp_shard,
+            )
+            self._pre_forward(self._handles, unused, unused, unused)
+            self._cast_forward_inputs(*args, **kwargs)
+            output = self._fsdp_wrapped_module(*args, **kwargs)
+            return self._post_forward(self._handles, reshard_fn, unused, unused, output)
+
+    def _register_pre_forward_hooks(self):
+        """
+        Registers pre-forward hooks to all modules in the wrapped root module's
+        hierarchy. The pre-forward hooks are partially applied based on the
+        current flattened parameter handle construction, which means that they
+        must be re-registered if the construction changes.
+        """
+        assert self._use_param_exec_order_policy
+        for forward_handle in self._pre_forward_handles:
+            forward_handle.remove()
+        self._pre_forward_handles.clear()
+        for module in self.module.modules():
+            module_flat_param_handles = self._module_to_handles[module]
```

Comment:
My understanding is that `self._module_to_handles[module]` contains all handles to flat parameters inside module.
If that is the case, line 2582-2584 is trying to reshard these parameters before the forward, which I am still a bit confused. I guess here we want to reshard the params of the previous module?

Reply:
Ah, I think I implemented it wrong. I understand the confusion 😅

Reply:
I think in my original PR (or at least in my mind), I used to pass all handles in, and the pre-foward reshard would check which ones are currently unsharded and reshard those.